File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>

class Solution {
public:
    bool validPath(int n, std::vector<std::vector<int>>& edges, int source, int destination) {
        if (destination == source) {
            return true;
        }

        std::unordered_map<int, std::vector<int>> graph;
        std::unordered_set<int> visited;

        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        return dfs(graph, visited, source, destination);
    }

private:
    bool dfs(const std::unordered_map<int, std::vector<int>>& graph, std::unordered_set<int>& visited, int node, int destination) {
        if (node == destination) {
            return true;
        }

        visited.insert(node);

        for (int neighbor : graph.at(node)) {
            if (visited.find(neighbor) == visited.end()) {
                if (dfs(graph, visited, neighbor, destination)) {
                    return true;
                }
            }
        }

        return false;
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
    if (destination === source) {
        return true;
    }

    const graph = new Map();
    const visited = new Set();

    for (const [u, v] of edges) {
        if (!graph.has(u)) {
            graph.set(u, []);
        }
        graph.get(u).push(v);

        if (!graph.has(v)) {
            graph.set(v, []);
        }
        graph.get(v).push(u);
    }

    return dfs(graph, visited, source, destination);
}

function dfs(graph, visited, node, destination) {
    if (node === destination) {
        return true;
    }

    visited.add(node);

    for (const neighbor of graph.get(node) || []) {
        if (!visited.has(neighbor)) {
            if (dfs(graph, visited, neighbor, destination)) {
                return true;
            }
        }
    }

    return false;
}
